src/main/java/io/cdap/plugin/gcp/spanner/common/SpannerUtil.java

```diff
@@ -38,12 +52,101 @@
+                                             Counter counter) throws IOException {
+    SpannerOptions.Builder optionsBuilder = buildSpannerOptions(serviceAccountFilePath, projectId);
+    optionsBuilder.setInterceptorProvider(
+      SpannerInterceptorProvider.createDefault()
+      .with(
+        new ClientInterceptor() {
+          @Override
+          public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+            return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+              next.newCall(method, callOptions)) {
+              @Override
+              public void start(Listener<RespT> responseListener, Metadata headers) {
+                super.start(new ForwardingClientCallListener.
+                  SimpleForwardingClientCallListener<RespT>(responseListener) {
+                  @Override
+                  public void onMessage(RespT message) {
+                    if (message instanceof PartialResultSet) {
+                      PartialResultSet partialResultSet = (PartialResultSet) message;
+                      counter.increment(partialResultSet.getSerializedSize());
+                    } else if (message instanceof ResultSet) {
+                      ResultSet resultSet = (ResultSet) message;
+                      counter.increment(resultSet.getSerializedSize());
+                    }
+                    super.onMessage(message);
+                  }
+                }, headers);
+              }
+            };
+          }
+        }
+      ));
+
+    return optionsBuilder.build().getService();
+  }
+
+  /**
+   * Construct and return the {@link Spanner} service with an interceptor that increments the provided counter by the
+   * number of bytes written to Spanner.
+   */
+  public static Spanner getSpannerServiceWithWriteInterceptor(String serviceAccountFilePath, String projectId,
+                                                             Counter counter) throws IOException {
+    SpannerOptions.Builder optionsBuilder = buildSpannerOptions(serviceAccountFilePath, projectId);
+    optionsBuilder.setInterceptorProvider(
+      SpannerInterceptorProvider.createDefault()
+        .with(
+          new ClientInterceptor() {
+            @Override
+            public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+              MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+              return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+                next.newCall(method, callOptions)) {
+                @Override
+                public void sendMessage(ReqT message) {
+                  if (message instanceof CommitRequest) {
+                    CommitRequest commitRequest = (CommitRequest) message;
+                    for (Mutation mutation: commitRequest.getMutationsList()) {
```

Comment:
Probably would be cleanly if using Java stream:

```java
    commitRequest.getMutationsList().stream()
      .map(GeneratedMessageV3::getAllFields)
      .map(Map::values)
      .flatMap(Collection::stream)
      .filter(Mutation.Write.class::isInstance)
      .map(Mutation.Write.class::cast)
      .map(Mutation.Write::getSerializedSize)
      .forEach(counter::increment);
```

Reply:
Thanks, this is much better! fixed.